File: go/graphics/bulk_picture.go
// Bulk Picture Manipulator — transformation pipeline applied over a batch.
package graphics

import "fmt"

type OpKind int

const (
	OpResize OpKind = iota
	OpGrayscale
	OpRotate90
	OpRotate180
	OpRotate270
	OpFlipH
	OpFlipV
	OpCrop
	OpBrightness
	OpContrast
)

type Op struct {
	Kind            OpKind
	Width, Height   uint32
	X, Y            uint32
	Algorithm       GSAlgorithm
	BrightnessDelta int
	ContrastFactor  float64
}

type OpError struct{ Reason string }

func (e *OpError) Error() string { return e.Reason }

type Pipeline []Op

func applyOp(img *GSImage, op Op) (*GSImage, error) {
	switch op.Kind {
	case OpResize:
		if op.Width == 0 || op.Height == 0 {
			return nil, &OpError{"zero_dimension"}
		}
		return resizeNearest(img, op.Width, op.Height), nil
	case OpGrayscale:
		return grayscaleImg(img, op.Algorithm), nil
	case OpRotate90:
		return rotate90(img), nil
	case OpRotate180:
		return rotate180(img), nil
	case OpRotate270:
		return rotate270(img), nil
	case OpFlipH:
		return flipH(img), nil
	case OpFlipV:
		return flipV(img), nil
	case OpCrop:
		if op.Width == 0 || op.Height == 0 {
			return nil, &OpError{"zero_dimension"}
		}
		if op.X+op.Width > img.Width || op.Y+op.Height > img.Height {
			return nil, &OpError{"crop_out_of_bounds"}
		}
		return cropImg(img, op.X, op.Y, op.Width, op.Height), nil
	case OpBrightness:
		return mapPixels(img, func(p GSRgb) GSRgb { return adjBrightness(p, op.BrightnessDelta) }), nil
	case OpContrast:
		return mapPixels(img, func(p GSRgb) GSRgb { return adjContrast(p, op.ContrastFactor) }), nil
	}
	return nil, &OpError{"unknown_op"}
}

func ApplyPipeline(img *GSImage, pipeline Pipeline) (*GSImage, error) {
	cur := img
	for _, op := range pipeline {
		next, err := applyOp(cur, op)
		if err != nil {
			return nil, err
		}
		cur = next
	}
	return cur, nil
}

type BatchResult struct {
	Index  int
	Output *GSImage
	Error  string
}

func RunBatch(images []*GSImage, pipeline Pipeline) []BatchResult {
	results := make([]BatchResult, 0, len(images))
	for i, img := range images {
		out, err := ApplyPipeline(img, pipeline)
		if err != nil {
			results = append(results, BatchResult{Index: i, Error: err.Error()})
		} else {
			results = append(results, BatchResult{Index: i, Output: out})
		}
	}
	return results
}

func DryRunDims(start [2]uint32, pipeline Pipeline) ([][2]uint32, error) {
	out := [][2]uint32{start}
	cur := start
	for _, op := range pipeline {
		switch op.Kind {
		case OpResize:
			if op.Width == 0 || op.Height == 0 {
				return nil, &OpError{"zero_dimension"}
			}
			cur = [2]uint32{op.Width, op.Height}
		case OpGrayscale, OpFlipH, OpFlipV, OpRotate180, OpBrightness, OpContrast:
			// unchanged
		case OpRotate90, OpRotate270:
			cur = [2]uint32{cur[1], cur[0]}
		case OpCrop:
			if op.Width == 0 || op.Height == 0 {
				return nil, &OpError{"zero_dimension"}
			}
			if op.X+op.Width > cur[0] || op.Y+op.Height > cur[1] {
				return nil, &OpError{"crop_out_of_bounds"}
			}
			cur = [2]uint32{op.Width, op.Height}
		}
		out = append(out, cur)
	}
	return out, nil
}

// ---- constructors ----

func OpResizeNew(w, h uint32) Op { return Op{Kind: OpResize, Width: w, Height: h} }
func OpGrayscaleNew(a GSAlgorithm) Op { return Op{Kind: OpGrayscale, Algorithm: a} }
func OpRotate90New() Op { return Op{Kind: OpRotate90} }
func OpRotate180New() Op { return Op{Kind: OpRotate180} }
func OpRotate270New() Op { return Op{Kind: OpRotate270} }
func OpFlipHNew() Op { return Op{Kind: OpFlipH} }
func OpFlipVNew() Op { return Op{Kind: OpFlipV} }
func OpCropNew(x, y, w, h uint32) Op {
	return Op{Kind: OpCrop, X: x, Y: y, Width: w, Height: h}
}
func OpBrightnessNew(d int) Op { return Op{Kind: OpBrightness, BrightnessDelta: d} }
func OpContrastNew(f float64) Op { return Op{Kind: OpContrast, ContrastFactor: f} }

// ---- primitives ----

func clampU8(v int) uint8 {
	if v < 0 { return 0 }
	if v > 255 { return 255 }
	return uint8(v)
}

func adjBrightness(p GSRgb, d int) GSRgb {
	return GSRgb{clampU8(int(p.R) + d), clampU8(int(p.G) + d), clampU8(int(p.B) + d)}
}

func adjContrast(p GSRgb, f float64) GSRgb {
	adj := func(v uint8) uint8 {
		x := (float64(v)-128.0)*f + 128.0
		return clampU8(int(round(x)))
	}
	return GSRgb{adj(p.R), adj(p.G), adj(p.B)}
}

func round(x float64) float64 {
	if x >= 0 { return float64(int64(x + 0.5)) }
	return float64(int64(x - 0.5))
}

func mapPixels(img *GSImage, f func(GSRgb) GSRgb) *GSImage {
	out := make([]GSRgb, len(img.Pixels))
	for i, p := range img.Pixels {
		out[i] = f(p)
	}
	return &GSImage{Width: img.Width, Height: img.Height, Pixels: out}
}

func grayscaleImg(img *GSImage, algo GSAlgorithm) *GSImage {
	return mapPixels(img, func(p GSRgb) GSRgb {
		v := GSPixelToGray(p, algo)
		return GSRgb{v, v, v}
	})
}

func resizeNearest(img *GSImage, nw, nh uint32) *GSImage {
	pixels := make([]GSRgb, int(nw)*int(nh))
	for y := uint32(0); y < nh; y++ {
		for x := uint32(0); x < nw; x++ {
			sx := uint32(uint64(x) * uint64(img.Width) / uint64(nw))
			sy := uint32(uint64(y) * uint64(img.Height) / uint64(nh))
			if sx >= img.Width { sx = img.Width - 1 }
			if sy >= img.Height { sy = img.Height - 1 }
			p, _ := img.Get(sx, sy)
			pixels[int(y)*int(nw)+int(x)] = p
		}
	}
	return &GSImage{Width: nw, Height: nh, Pixels: pixels}
}

func buildImage(w, h uint32, cell func(x, y uint32) GSRgb) *GSImage {
	pixels := make([]GSRgb, int(w)*int(h))
	for y := uint32(0); y < h; y++ {
		for x := uint32(0); x < w; x++ {
			pixels[int(y)*int(w)+int(x)] = cell(x, y)
		}
	}
	return &GSImage{Width: w, Height: h, Pixels: pixels}
}

func rotate90(img *GSImage) *GSImage {
	w, h := img.Width, img.Height
	return buildImage(h, w, func(nx, ny uint32) GSRgb {
		p, _ := img.Get(ny, h-1-nx)
		return p
	})
}

func rotate180(img *GSImage) *GSImage {
	w, h := img.Width, img.Height
	return buildImage(w, h, func(nx, ny uint32) GSRgb {
		p, _ := img.Get(w-1-nx, h-1-ny)
		return p
	})
}

func rotate270(img *GSImage) *GSImage {
	w, h := img.Width, img.Height
	return buildImage(h, w, func(nx, ny uint32) GSRgb {
		p, _ := img.Get(w-1-ny, nx)
		return p
	})
}

func flipH(img *GSImage) *GSImage {
	w, h := img.Width, img.Height
	return buildImage(w, h, func(nx, ny uint32) GSRgb {
		p, _ := img.Get(w-1-nx, ny)
		return p
	})
}

func flipV(img *GSImage) *GSImage {
	w, h := img.Width, img.Height
	return buildImage(w, h, func(nx, ny uint32) GSRgb {
		p, _ := img.Get(nx, h-1-ny)
		return p
	})
}

func cropImg(img *GSImage, x0, y0, w, h uint32) *GSImage {
	return buildImage(w, h, func(nx, ny uint32) GSRgb {
		p, _ := img.Get(x0+nx, y0+ny)
		return p
	})
}

func BulkPictureDemo() {
	pixels := make([]GSRgb, 16)
	for y := 0; y < 4; y++ {
		for x := 0; x < 4; x++ {
			if (x+y)%2 == 0 {
				pixels[y*4+x] = GSRgb{200, 100, 50}
			} else {
				pixels[y*4+x] = GSRgb{50, 100, 200}
			}
		}
	}
	img := &GSImage{Width: 4, Height: 4, Pixels: pixels}

	pipeline := Pipeline{
		OpRotate90New(),
		OpBrightnessNew(10),
		OpGrayscaleNew(GSLuminosity),
	}

	out, _ := ApplyPipeline(img, pipeline)
	dims, _ := DryRunDims([2]uint32{4, 4}, pipeline)
	fmt.Printf("pipeline steps: %d\n", len(pipeline))
	fmt.Printf("dry-run dims: %v\n", dims)
	p := out.Pixels[0]
	fmt.Printf("output: %dx%d pixels[0]=(%d,%d,%d)\n", out.Width, out.Height, p.R, p.G, p.B)

	batch := []*GSImage{
		GSImageFilled(6, 6, GSRgb{128, 128, 128}),
		GSImageFilled(2, 2, GSRgb{128, 128, 128}),
	}
	cropPipe := Pipeline{OpCropNew(0, 0, 3, 3)}
	for _, r := range RunBatch(batch, cropPipe) {
		if r.Output != nil {
			fmt.Printf("  [%d] ok: %dx%d\n", r.Index, r.Output.Width, r.Output.Height)
		} else {
			fmt.Printf("  [%d] err: %s\n", r.Index, r.Error)
		}
	}
}

File: python/graphics/bulk_picture.py
"""Bulk Picture Manipulator — transformation pipeline applied over a batch.

Pipeline is data: Op union + list of ops. Pure fold. Dry-run predicts
dims without touching pixels. Batch isolates per-image errors.
"""

from __future__ import annotations
from dataclasses import dataclass
from enum import Enum
from typing import Union

from graphics.grayscale import Algorithm, Image, Rgb, pixel_to_gray


class OpKind(Enum):
    RESIZE = "resize"
    GRAYSCALE = "grayscale"
    ROTATE_90 = "rotate_90"
    ROTATE_180 = "rotate_180"
    ROTATE_270 = "rotate_270"
    FLIP_H = "flip_h"
    FLIP_V = "flip_v"
    CROP = "crop"
    BRIGHTNESS = "brightness"
    CONTRAST = "contrast"


@dataclass(frozen=True)
class Op:
    kind: OpKind
    # payload fields — only the relevant ones are populated per kind
    width: int = 0
    height: int = 0
    x: int = 0
    y: int = 0
    algorithm: Algorithm = Algorithm.LUMINOSITY
    brightness_delta: int = 0
    contrast_factor: float = 1.0


def op_resize(w: int, h: int) -> Op:
    return Op(OpKind.RESIZE, width=w, height=h)

def op_grayscale(algo: Algorithm) -> Op:
    return Op(OpKind.GRAYSCALE, algorithm=algo)

def op_rotate_90() -> Op: return Op(OpKind.ROTATE_90)
def op_rotate_180() -> Op: return Op(OpKind.ROTATE_180)
def op_rotate_270() -> Op: return Op(OpKind.ROTATE_270)
def op_flip_h() -> Op: return Op(OpKind.FLIP_H)
def op_flip_v() -> Op: return Op(OpKind.FLIP_V)

def op_crop(x: int, y: int, w: int, h: int) -> Op:
    return Op(OpKind.CROP, x=x, y=y, width=w, height=h)

def op_brightness(delta: int) -> Op:
    return Op(OpKind.BRIGHTNESS, brightness_delta=delta)

def op_contrast(factor: float) -> Op:
    return Op(OpKind.CONTRAST, contrast_factor=factor)


class OpError(Exception):
    pass


class CropOutOfBounds(OpError):
    pass


class ZeroDimension(OpError):
    pass


Pipeline = list[Op]


def apply_op(img: Image, op: Op) -> Image:
    k = op.kind
    if k == OpKind.RESIZE:
        if op.width == 0 or op.height == 0: raise ZeroDimension()
        return _resize_nearest(img, op.width, op.height)
    if k == OpKind.GRAYSCALE:
        return _grayscale(img, op.algorithm)
    if k == OpKind.ROTATE_90:
        return _rotate_90(img)
    if k == OpKind.ROTATE_180:
        return _rotate_180(img)
    if k == OpKind.ROTATE_270:
        return _rotate_270(img)
    if k == OpKind.FLIP_H:
        return _flip_h(img)
    if k == OpKind.FLIP_V:
        return _flip_v(img)
    if k == OpKind.CROP:
        if op.width == 0 or op.height == 0: raise ZeroDimension()
        if op.x + op.width > img.width or op.y + op.height > img.height:
            raise CropOutOfBounds()
        return _crop(img, op.x, op.y, op.width, op.height)
    if k == OpKind.BRIGHTNESS:
        return _map_pixels(img, lambda p: _adj_brightness(p, op.brightness_delta))
    if k == OpKind.CONTRAST:
        return _map_pixels(img, lambda p: _adj_contrast(p, op.contrast_factor))
    raise AssertionError(f"unreachable: {k}")


def apply_pipeline(img: Image, pipeline: Pipeline) -> Image:
    cur = img
    for op in pipeline:
        cur = apply_op(cur, op)
    return cur


@dataclass
class BatchOk:
    index: int
    output: Image

@dataclass
class BatchErr:
    index: int
    error: str

BatchResult = Union[BatchOk, BatchErr]


def run_batch(images: list[Image], pipeline: Pipeline) -> list[BatchResult]:
    results: list[BatchResult] = []
    for i, img in enumerate(images):
        try:
            out = apply_pipeline(img, pipeline)
            results.append(BatchOk(i, out))
        except OpError as e:
            results.append(BatchErr(i, type(e).__name__))
    return results


def dry_run_dims(start: tuple[int, int], pipeline: Pipeline) -> list[tuple[int, int]]:
    out = [start]
    cur = start
    for op in pipeline:
        k = op.kind
        if k == OpKind.RESIZE:
            if op.width == 0 or op.height == 0: raise ZeroDimension()
            cur = (op.width, op.height)
        elif k in (OpKind.GRAYSCALE, OpKind.FLIP_H, OpKind.FLIP_V,
                   OpKind.ROTATE_180, OpKind.BRIGHTNESS, OpKind.CONTRAST):
            pass  # cur unchanged
        elif k in (OpKind.ROTATE_90, OpKind.ROTATE_270):
            cur = (cur[1], cur[0])
        elif k == OpKind.CROP:
            if op.width == 0 or op.height == 0: raise ZeroDimension()
            if op.x + op.width > cur[0] or op.y + op.height > cur[1]:
                raise CropOutOfBounds()
            cur = (op.width, op.height)
        out.append(cur)
    return out


# ---------- primitives ----------

def _map_pixels(img: Image, f) -> Image:
    return Image(img.width, img.height, [f(p) for p in img.pixels])


def _adj_brightness(p: Rgb, delta: int) -> Rgb:
    clamp = lambda v: max(0, min(255, v + delta))
    return Rgb(clamp(p.r), clamp(p.g), clamp(p.b))


def _adj_contrast(p: Rgb, factor: float) -> Rgb:
    def adj(v: int) -> int:
        return max(0, min(255, round((v - 128) * factor + 128)))
    return Rgb(adj(p.r), adj(p.g), adj(p.b))


def _grayscale(img: Image, algo: Algorithm) -> Image:
    def to_gray_px(p: Rgb) -> Rgb:
        v = pixel_to_gray(p, algo)
        return Rgb(v, v, v)
    return _map_pixels(img, to_gray_px)


def _get(img: Image, x: int, y: int) -> Rgb:
    return img.pixels[y * img.width + x]


def _resize_nearest(img: Image, new_w: int, new_h: int) -> Image:
    pixels: list[Rgb] = []
    for y in range(new_h):
        for x in range(new_w):
            sx = min(img.width - 1, (x * img.width) // new_w)
            sy = min(img.height - 1, (y * img.height) // new_h)
            pixels.append(_get(img, sx, sy))
    return Image(new_w, new_h, pixels)


def _rotate_90(img: Image) -> Image:
    new = [Rgb(0, 0, 0)] * (img.width * img.height)
    for y in range(img.height):
        for x in range(img.width):
            new_x = img.height - 1 - y
            new_y = x
            new[new_y * img.height + new_x] = _get(img, x, y)
    return Image(img.height, img.width, new)


def _rotate_180(img: Image) -> Image:
    return Image(img.width, img.height, list(reversed(img.pixels)))


def _rotate_270(img: Image) -> Image:
    new = [Rgb(0, 0, 0)] * (img.width * img.height)
    for y in range(img.height):
        for x in range(img.width):
            new_x = y
            new_y = img.width - 1 - x
            new[new_y * img.height + new_x] = _get(img, x, y)
    return Image(img.height, img.width, new)


def _flip_h(img: Image) -> Image:
    new = [Rgb(0, 0, 0)] * (img.width * img.height)
    for y in range(img.height):
        for x in range(img.width):
            new[y * img.width + (img.width - 1 - x)] = _get(img, x, y)
    return Image(img.width, img.height, new)


def _flip_v(img: Image) -> Image:
    new = [Rgb(0, 0, 0)] * (img.width * img.height)
    for y in range(img.height):
        for x in range(img.width):
            new[(img.height - 1 - y) * img.width + x] = _get(img, x, y)
    return Image(img.width, img.height, new)


def _crop(img: Image, x0: int, y0: int, w: int, h: int) -> Image:
    pixels: list[Rgb] = []
    for y in range(h):
        for x in range(w):
            pixels.append(_get(img, x0 + x, y0 + y))
    return Image(w, h, pixels)


def demo() -> None:
    # 4x4 checker
    pixels = []
    for y in range(4):
        for x in range(4):
            if (x + y) % 2 == 0:
                pixels.append(Rgb(200, 100, 50))
            else:
                pixels.append(Rgb(50, 100, 200))
    img = Image(4, 4, pixels)

    pipeline: Pipeline = [
        op_rotate_90(),
        op_brightness(10),
        op_grayscale(Algorithm.LUMINOSITY),
    ]

    out = apply_pipeline(img, pipeline)
    dims = dry_run_dims((4, 4), pipeline)
    print(f"pipeline steps: {len(pipeline)}")
    print(f"dry-run dims: {dims}")
    print(f"output: {out.width}x{out.height} pixels[0]=({out.pixels[0].r},{out.pixels[0].g},{out.pixels[0].b})")

    # batch with a mix of valid and invalid
    batch = [Image(6, 6, [Rgb(128, 128, 128)] * 36),
             Image(2, 2, [Rgb(128, 128, 128)] * 4)]
    crop_pipeline = [op_crop(0, 0, 3, 3)]
    results = run_batch(batch, crop_pipeline)
    for r in results:
        if isinstance(r, BatchOk):
            print(f"  [{r.index}] ok: {r.output.width}x{r.output.height}")
        else:
            print(f"  [{r.index}] err: {r.error}")


if __name__ == "__main__":
    demo()
